NeoML/src/TraditionalML/SparseFloatMatrix.cpp

```diff
@@ -21,96 +21,112 @@ limitations under the License.
 namespace NeoML {
 
 CFloatMatrixDesc CFloatMatrixDesc::Empty;
-
-CSparseFloatMatrix::CSparseFloatMatrixBody* CSparseFloatMatrix::CSparseFloatMatrixBody::Duplicate() const
-{
-	CSparseFloatMatrixBody* body = FINE_DEBUG_NEW CSparseFloatMatrixBody( Desc.Height, Desc.Width, ElementCount, RowsBufferSize, ElementsBufferSize );
-	::memcpy( body->Desc.Columns, Desc.Columns, ElementCount * sizeof( int ) );
-	::memcpy( body->Desc.Values, Desc.Values, ElementCount * sizeof( float ) );
-	::memcpy( body->Desc.PointerB, Desc.PointerB, Desc.Height * sizeof( int ) );
-	::memcpy( body->Desc.PointerE, Desc.PointerE, Desc.Height * sizeof( int ) );
-	return body;
-}
+static const int InitialRowsBufferSize = 32;
+static const int InitialElementsBufferSize = 512;
+static const int MaxBufferSize = INT_MAX;
 
 CSparseFloatMatrix::CSparseFloatMatrixBody::CSparseFloatMatrixBody( int height, int width, int elementCount,
 		int rowsBufferSize, int elementsBufferSize ) :
 	RowsBufferSize( rowsBufferSize ),
 	ElementsBufferSize( elementsBufferSize ),
 	ElementCount( elementCount )
 {
-	NeoAssert( RowsBufferSize >= 0 );
-	NeoAssert( ElementsBufferSize >= 0 );
+	NeoAssert( height >= 0 && width >= 0 && elementCount >= 0 );
+	NeoAssert( rowsBufferSize >= 0 && elementsBufferSize >= 0 );
+
 	Desc.Height = height;
 	Desc.Width = width;
 
-	ColumnsBuf.SetSize( ElementsBufferSize );
-	ValuesBuf.SetSize( ElementsBufferSize );
-	BeginPointersBuf.SetSize( RowsBufferSize );
-	EndPointersBuf.SetSize( RowsBufferSize );
-
-	Desc.Columns = ColumnsBuf.GetPtr();
-	Desc.Values = ValuesBuf.GetPtr();
-	Desc.PointerB = BeginPointersBuf.GetPtr();
-	Desc.PointerE = EndPointersBuf.GetPtr();
+	if( ElementsBufferSize > 0 ) {
+		ElementsBufferSize = max( ElementsBufferSize, InitialElementsBufferSize );
+		Desc.Columns = FINE_DEBUG_NEW int[ElementsBufferSize];
+		Desc.Values = FINE_DEBUG_NEW float[ElementsBufferSize];
+	}
+	if( RowsBufferSize > 0 ) {
+		RowsBufferSize = max( RowsBufferSize, InitialRowsBufferSize );
```

Comment:
PointerB/E should have at least Desc.Height size!

Reply:
added NeoAssert for this